nexus/db-queries/src/db/datastore/sled.rs

```diff
@@ -217,6 +232,34 @@ impl DataStore {
             .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))?;
         Ok(())
     }
+
+    /// Sets the provision state for this sled.
+    pub async fn sled_set_provision_state(
+        &self,
+        opctx: &OpContext,
+        sled_lookup: &lookup::Sled<'_>,
+        state: db::model::SledProvisionState,
+    ) -> Result<db::model::SledProvisionState, external::Error> {
+        // Ensure that opctx is authorized to modify this sled.
+        let (_, sled) = sled_lookup.fetch_for(authz::Action::Modify).await?;
+        let old_state = sled.provision_state();
+        if old_state == state {
+            return Ok(old_state);
+        }
+
+        use db::schema::sled::dsl;
+        // XXX: Do we need transaction semantics here? Probably not because
+        // this is idempotent?
```

Comment:
There are a couple very small tweaks I might make here:

Add `filter(dsl::time_deleted.is_null())` when searching for this sled. If we've decided the sled should be deleted, we should not enable further modifications.

On transactions: This is a single statement, so it is implicitly treated as a transaction: https://www.cockroachlabs.com/docs/stable/transactions#individual-statements . The only part of this that *isn't* considered transactional is that in-memory check of "if old_state == state" that's being done in memory, because someone could modify the DB state while we're making that state.

This is a very small concern, but I'll mention it because I think it's a good example for cases that could be more complex:

Suppose we have multiple operators trying to update provision state from "provisionable" to "not provisionable" at the same time. If both get past the `if old_state == state` check before getting to the `UPDATE`, then they'll both update the sled state (to the same value!), and both see the `old_state` return value of "provisionable", even though (according to a serialized view of the world) "one of them actually made the sled "not provisionable".

(I don't actually think this is that bad, but it *does* raise the question of "what does that return value mean? **should** we actually be exposing that information? How much do we care about providing genuine "compare-and-swap" semantics here?)

This is one of the reasons why it's kinda nice to perform as many checks as we'd like within "purely SQL" -- the single statement is an implicit transaction, so we avoid races like this by doing this kinda work in memory.

- If we add a `dsl::sled.filter(dsl::provision_state.ne(state))` clause to the `UPDATE`, then we'll only perform this update if we actually have work to do, by extending the `WHERE` clause. However, this has a downside -- if the sled **exists**, but is already in the desired state, it'll be filtered out, and we'll see "no rows updated".
- We can use use the `RETURNING` clause to access the DB value, but this only returns the **new** value, and not the old one.

Fortunately, there is a CTE that we have built to help queries do the following: https://github.com/oxidecomputer/omicron/blob/main/nexus/db-queries/src/db/update_and_check.rs

This lets you...
- Filter for the exact condition you want to check
- ... but if the old object exists, you can read it from the DB: https://github.com/oxidecomputer/omicron/blob/55b39533cfe9a3f2fc1185adaa9c2118efaee6bf/nexus/db-queries/src/db/update_and_check.rs#L118-L123

This basically lets you do an update **and** read the old value at the same time, if you want. Look for `check_if_exists` if you wanna find some example usage within the codebase.

That said -- we only really want this if we want to return the old state value! If we simply want to omit that from the return value, I think we could mostly ignore this machinery.

Reply:
Thank you, the CTE is exactly what I was looking for, I think!

Reply:
All right, I updated the PR to use `update_and_check` but I'm seeing a mysterious error: https://gist.github.com/sunshowers/5e0829d0a29f09ae03f5782fc094206f#file-gistfile0-txt-L28

I verified that it is the `update_and_check` that's causing this by commenting it out and seeing these tests pass.

Reply:
This was https://github.com/oxidecomputer/omicron/issues/4570.

Reply:
Addressed using `update_and_check`, works great.